nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java

```diff
@@ -0,0 +1,334 @@
+import org.apache.iceberg.util.Tasks;
+import org.projectnessie.client.NessieConfigConstants;
+import org.projectnessie.client.api.CommitMultipleOperationsBuilder;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.client.http.HttpClientException;
+import org.projectnessie.error.BaseNessieClientServerException;
+import org.projectnessie.error.NessieConflictException;
+import org.projectnessie.error.NessieNamespaceAlreadyExistsException;
+import org.projectnessie.error.NessieNamespaceNotEmptyException;
+import org.projectnessie.error.NessieNamespaceNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.error.NessieReferenceNotFoundException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.EntriesResponse;
+import org.projectnessie.model.GetNamespacesResponse;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.Operation;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.Tag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NessieIcebergClient implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(NessieIcebergClient.class);
+
+  private final NessieApiV1 api;
+  private final Supplier<UpdateableReference> reference;
+  private final Map<String, String> catalogOptions;
+
+  public NessieIcebergClient(
+      NessieApiV1 api, String requestedRef, String requestedHash, Map<String, String> catalogOptions) {
+    this.api = api;
+    this.catalogOptions = catalogOptions;
+    this.reference = () -> loadReference(requestedRef, requestedHash);
+  }
+
+  public NessieApiV1 getApi() {
+    return api;
+  }
+
+  public UpdateableReference getRef() {
+    return reference.get();
+  }
+
+  public void refresh() throws NessieNotFoundException {
+    getRef().refresh(api);
+  }
+
+  public NessieIcebergClient withReference(String requestedRef, String hash) {
+    if (null == requestedRef) {
+      return this;
+    }
+    return new NessieIcebergClient(getApi(), requestedRef, hash, catalogOptions);
+  }
+
+  private UpdateableReference loadReference(String requestedRef, String hash) {
+    try {
+      Reference ref =
+          requestedRef == null ? api.getDefaultBranch() : api.getReference().refName(requestedRef).get();
+      if (hash != null) {
+        if (ref instanceof Branch) {
+          ref = Branch.of(ref.getName(), hash);
+        } else {
+          ref = Tag.of(ref.getName(), hash);
+        }
+      }
+      return new UpdateableReference(ref, hash != null);
+    } catch (NessieNotFoundException ex) {
+      if (requestedRef != null) {
+        throw new IllegalArgumentException(String.format("Nessie ref '%s' does not exist. This ref must exist " +
+            "before creating a NessieCatalog.", requestedRef), ex);
```

Comment:
Message here may be confusing since this is no longer part of a catalog class, although it only would be called by a catalog class ....

Reply:
updated